gateway/blocks_gateway.go

```diff
@@ -223,38 +231,215 @@ func (api *BlocksGateway) Head(ctx context.Context, path ImmutablePath) (Content
 	return md, fileNode, nil
 }
 
-func (api *BlocksGateway) GetCAR(ctx context.Context, path ImmutablePath) (ContentPathMetadata, io.ReadCloser, <-chan error, error) {
-	// Same go-car settings as dag.export command
-	store := dagStore{api: api, ctx: ctx}
-
-	// TODO: When switching to exposing path blocks we'll want to add these as well
-	roots, lastSeg, err := api.getPathRoots(ctx, path)
-	if err != nil {
-		return ContentPathMetadata{}, nil, nil, err
+func (api *BlocksGateway) GetCAR(ctx context.Context, p ImmutablePath, params CarParams) (io.ReadCloser, <-chan error, error) {
+	contentPathStr := p.String()
+	if !strings.HasPrefix(contentPathStr, "/ipfs/") {
+		return nil, nil, fmt.Errorf("path does not have /ipfs/ prefix")
 	}
-
-	md := ContentPathMetadata{
-		PathSegmentRoots: roots,
-		LastSegment:      lastSeg,
+	firstSegment, _, _ := strings.Cut(contentPathStr[6:], "/")
+	rootCid, err := cid.Decode(firstSegment)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	rootCid := lastSeg.Cid()
-
-	// TODO: support selectors passed as request param: https://github.com/ipfs/kubo/issues/8769
-	// TODO: this is very slow if blocks are remote due to linear traversal. Do we need deterministic traversals here?
-	dag := car.Dag{Root: rootCid, Selector: selectorparse.CommonSelector_ExploreAllRecursively}
-	c := car.NewSelectiveCar(ctx, store, []car.Dag{dag}, car.TraverseLinksOnlyOnce())
 	r, w := io.Pipe()
-
 	errCh := make(chan error, 1)
 	go func() {
-		carWriteErr := c.Write(w)
+		cw, err := storage.NewWritable(w, []cid.Cid{rootCid}, car.WriteAsCarV1(true))
+		if err != nil {
+			pipeCloseErr := w.Close()
+			errCh <- multierr.Combine(err, pipeCloseErr)
+			close(errCh)
+			return
+		}
+
+		blockGetter := merkledag.NewDAGService(api.blockService).Session(ctx)
+
+		blockGetter = &nodeGetterToCarExporer{
+			ng: blockGetter,
+			cw: cw,
+		}
+
+		// Setup the UnixFS resolver.
+		f := newNodeGetterFetcherSingleUseFactory(ctx, blockGetter)
```

Comment:
This new struct is an unfortunate piece of glue. AFAIK the only fetcher implementation that exists is https://github.com/ipfs/go-fetcher/blob/99235cc120519afe2bddcb6b1f7a6844ecc3707f/impl/blockservice/fetcher.go#L32.

This is unfortunate as it means we can't use `ResolveToLastNode` anywhere which is needed because we have no mechanisms in go-ipld-prime/go-unixfsnode to figure out if the thing we got at the end of the path is UnixFS or not.

In short what we're dealing with is that the abstractions here make it easy to generically work with different types of data but if I want to figure out what the data is once we're in the abstraction (e.g. to check for HTTP Gateway API spec compliance) things become at least one of difficult or brittle.

Depending on feedback could either leave this code here, extract it to `go-fetcher` (which could also be copied into boxo), or extract the relevant code in the `resolver` package to not depend on `go-fetcher`

Reply:
I was also [wondering the same](https://github.com/ipfs/boxo/pull/303#discussion_r1213007373). Anyone has ideas here?

Reply:
I was also [wondering the same](https://github.com/ipfs/boxo/pull/303#discussion_r1213007373). Anyone has ideas here?